ratingcurve/ratingmodel_builder.py

```diff
@@ -0,0 +1,503 @@
+            Specifies whether the fit progressbar should be displayed.
+        random_seed : RandomState, optional
+            Provides sampler with initial random seed for obtaining
+            reproducible samples.
+        **kwargs : dict
+            Algorithm settings can be provided in form of keyword arguments.
+
+        Returns
+        -------
+        self : InferenceData
+            Arviz InferenceData object containing posterior samples of model
+            parameters of the fitted model.
+        """
+        # Save the fitting algorithm method
+        self.method = method
+
+        # Define the sampler configuration
+        sampler_config = self.get_default_sampler_config(**kwargs)
+        self.sampler_config = sampler_config
+
+        # Add progressbar and random_seed kwargs to sampler configuration
+        sampler_config = self.sampler_config.copy()
+        sampler_config["progressbar"] = progressbar
+        sampler_config["random_seed"] = random_seed
+        sampler_config.update(**kwargs)
+
+        # Build rating curve models which can include discharge uncertainty
+        # XXX This is where I expect generate and preprocess call
+        self.build_model(h, q, q_sigma=q_sigma)
+
+        # Sample (fit) the model
+        self.idata = self.sample_model(**sampler_config)
+
+        # Have fit data include uncertainty and have appropriate names
+        h_df = pd.DataFrame({'h': h})
+        q_df = pd.DataFrame({self.output_var: q})
+        if q_sigma is None:
+            q_sigma_df = pd.DataFrame({'q_sigma': np.zeros(len(q))})
+        else:
+            q_sigma_df = pd.DataFrame({'q_sigma': q_sigma})
+        combined_data = pd.concat([h_df, q_df, q_sigma_df], axis=1)
+
+        assert all(combined_data.columns), "All columns must have " + \
+                                           "non-empty names"
+        with warnings.catch_warnings():
+            warnings.filterwarnings(
+                "ignore",
+                category=UserWarning,
+                message="The group fit_data is not defined in "
+                        "the InferenceData scheme",
+            )
+            self.idata.add_groups(fit_data=combined_data.to_xarray())
+
+        return self.idata  # type: ignore
+
+    def sample_prior_predictive(self,
+                                X_pred,
+                                y_pred=None,
+                                samples: int = None,
+                                extend_idata: bool = False,
+                                combined: bool = True,
+                                **kwargs,
+                                ):
+        """Update of ModelBuilder `sample_prior_predicitve` function to
+        output untransformed discharge (q).
+        """
+        return self.q_transform.untransform(super().sample_prior_predictive(
+                  X_pred, y_pred, samples, extend_idata, combined, **kwargs))
+
+    def sample_posterior_predictive(self,
```

Comment:
both this function and `sample_posterior_predictive` are difficult to follow.
I think it would be clearer if you copied `model_builder.ModelBuilder.sample_posterior_predictive` rather than calling it through `super()`